libgraph/include/katana/GraphTopology.h

```diff
@@ -43,6 +45,59 @@ struct KATANA_EXPORT GraphTopologyTypes {
   using EdgeDestVec = NUMAArray<Node>;
   using PropIndexVec = NUMAArray<PropertyIndex>;
   using EntityTypeIDVec = NUMAArray<EntityTypeID>;
+  */
+
+  struct Node: public OpaqueIDLinear<Node, uint32_t> {
+    using OpaqueIDLinear::OpaqueIDLinear;
+  };
+
+  struct Edge: public OpaqueIDLinear<Edge, uint64_t> {
+    using OpaqueIDLinear::OpaqueIDLinear;
+  };
+
+  struct OutEdgeHandle: public OpaqueIDLinear<OutEdgeHandle, uint64_t> {
+    using OpaqueIDLinear::OpaqueIDLinear;
+  };
+
+  struct InEdgeHandle: public OpaqueIDLinear<InEdgeHandle, uint64_t> {
+    using OpaqueIDLinear::OpaqueIDLinear;
+  };
+
+  struct InOutEdgeHandle: public OpaqueIDLinear<InOutEdgeHandle, uint64_t> {
+    using OpaqueIDLinear::OpaqueIDLinear;
+  };
+
+  // TODO(Amber/emcginnis): Can this be uint32_t?
+  struct NodePropertyIndex: public OpaqueIDLinear<NodePropertyIndex, uint64_t> {
+    using OpaqueIDLinear::OpaqueIDLinear;
+  };
+
+  struct EdgePropertyIndex: public OpaqueIDLinear<EdgePropertyIndex, uint64_t> {
+    using OpaqueIDLinear::OpaqueIDLinear;
+  };
+
+  // TODO(Amber): Opaque IDs should define a separate iterator type instead of
+  // mixing the concepts in one class. 
+  using node_iterator = OpaqueIDCountingIterator<Node>;
+  using nodes_range = StandardRange<node_iterator>;
+
+  using edge_iterator = OpaqueIDCountingIterator<OutEdgeHandle>;
+  using edges_range = StandardRange<edge_iterator>;
+  // TODO(Amber): Remove because we have Nodes() method to iterate over all nodes
+  using iterator = node_iterator;
+
+  //TODO(emcginnis): Each of these *Vec types should really be *Array since they are not resizable
+  // NOTE(Amber): using underlying type so that (1) tusba & gluon partitioner interface doesn't
```

Comment:
Ah, so that's how you escaped the extra pain :-)

Reply:
There is an added benefit too. `std::vector` has to call a constructor for the entire vector when you create one with non-zero size. This construction is serial and can become a hidden time sink (and performance issue). So, we should be careful constructing large vectors of opaque types. `NUMAArray` on the other hand has a `constructAt` for this very reason (and can be invoked in a parallel `do_all`)